jet.py

```diff
@@ -124,7 +167,32 @@ def getP4Variations(self, df, source_dict, apply_JER=True, apply_JES=True):
         return df,source_dict
 
 
+    def getP4Variations_run3(self, df, source_dict, apply_JER, apply_JES):
+        df = df.Define("Jet_p4_shifted_map", f'''::correction::JetCorrectionProvider::getGlobal().getShiftedP4(Jet_pt, Jet_eta, Jet_phi, Jet_mass,
+                                                                                                               Jet_rawFactor, Jet_area, Rho_fixedGridRhoFastjetAll,
+                                                                                                               GenJet_pt, Jet_genJetIdx, event)''')
+
+        for source in JetCorrProducer.uncSources_core:
+            if source not in ["Central", "Total", "JER"]:
+                continue
+
+            updateSourceDict(source_dict, source, 'Jet')
+            for scale in getScales(source):
+                syst_name = getSystName(source, scale)
+
+                df = df.Define(f"Jet_p4_{syst_name}", f"Jet_p4_shifted_map.at({{::correction::JetCorrectionProvider::UncSource::{source}, ::correction::UncScale::{scale}}})")
+                df = df.Define(f"Jet_p4_{syst_name}_delta", f"Jet_p4_{syst_name} - Jet_p4_{nano}")
+
+        return df, source_dict
+
+
+    def getP4Variations(self, df, source_dict, apply_JER=True, apply_JES=True):
+        if self.year < 2022:
+            return self.getP4Variations_run2(df, source_dict, apply_JER, apply_JES)
+        else:
+            return self.getP4Variations_run3(df, source_dict, apply_JER, apply_JES)
+
+
     def getEnergyResolution(self, df):
-        df= df.Define(f"Jet_ptRes", f""" ::correction::JetCorrProvider::getGlobal().getResolution(
-            Jet_pt, Jet_eta, Rho_fixedGridRhoFastjetAll ) """)
-        return df
+        df= df.Define(f"Jet_ptRes", f""" ::correction::JetCorrProvider::getGlobal().getResolution(Jet_pt, Jet_eta, Rho_fixedGridRhoFastjetAll ) """)
```

Comment:
Make this function work also for corrlib case -> we would want to store jet resolution and use it for DNN/HME